packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart

```diff
@@ -0,0 +1,852 @@
+    goldenFile.createSync(recursive: true);
+    goldenFile.writeAsStringSync('''
+loading-units:
+  - id: 3
+    libraries:
+      - lib2
+      - lib3
+''', flush: true, mode: FileMode.append);
+    validator.checkAgainstLoadingUnitGolden(
+      <LoadingUnit>[
+        LoadingUnit(id: 2, libraries: <String>['lib1']),
+        LoadingUnit(id: 3, libraries: <String>['lib2', 'lib3']),
+      ],
+    );
+    validator.displayResults();
+    validator.attemptToolExit();
+
+    expect(logger.statusText.contains('New loading units were found:\n\n  LoadingUnit 2\n    Libraries:\n    - lib1\n'), true);
+  });
+
+  test('loadingUnitGolden finds missing loading units', () async {
+    final DeferredComponentsSetupValidator validator = DeferredComponentsSetupValidator(
+      env,
+      exitOnFail: false,
+      title: 'test check',
+    );
+    final File goldenFile = env.projectDir.childFile(DeferredComponentsSetupValidator.kDeferredComponentsGoldenFileName);
+    if (goldenFile.existsSync()) {
+      goldenFile.deleteSync();
+    }
+    goldenFile.createSync(recursive: true);
+    goldenFile.writeAsStringSync('''
+loading-units:
+  - id: 2
+    libraries:
+      - lib1
+  - id: 3
+    libraries:
+      - lib2
+      - lib3
+''', flush: true, mode: FileMode.append);
+    validator.checkAgainstLoadingUnitGolden(
+      <LoadingUnit>[
+        LoadingUnit(id: 3, libraries: <String>['lib2', 'lib3']),
+      ],
+    );
+    validator.displayResults();
+    validator.attemptToolExit();
+
+    expect(logger.statusText.contains('Previously existing loading units no longer exist:\n\n  LoadingUnit 2\n    Libraries:\n    - lib1\n'), true);
+  });
+
+  test('missing golden file counts as all new loading units', () async {
+    final DeferredComponentsSetupValidator validator = DeferredComponentsSetupValidator(
+      env,
+      exitOnFail: false,
+      title: 'test check',
+    );
+    final File goldenFile = env.projectDir.childFile(DeferredComponentsSetupValidator.kDeferredComponentsGoldenFileName);
+    if (goldenFile.existsSync()) {
+      goldenFile.deleteSync();
+    }
+    validator.checkAgainstLoadingUnitGolden(
+      <LoadingUnit>[
+        LoadingUnit(id: 2, libraries: <String>['lib1']),
+      ],
+    );
+    validator.displayResults();
+    validator.attemptToolExit();
+
+    expect(logger.statusText.contains('New loading units were found:\n\n  LoadingUnit 2\n    Libraries:\n    - lib1\n'), true);
+  });
+
+  test('androidRootSetup exists passes', () async {
+    final DeferredComponentsSetupValidator validator = DeferredComponentsSetupValidator(
+      env,
+      exitOnFail: false,
+      title: 'test check',
+    );
+    final File settingsGradleFile = env.projectDir.childDirectory('android').childFile('settings.gradle');
+    if (settingsGradleFile.existsSync()) {
+      settingsGradleFile.deleteSync();
+    }
+    settingsGradleFile.createSync(recursive: true);
+    settingsGradleFile.writeAsStringSync(r'''
+include ':app', ':component1'
```

Comment:
Same idea here. If `settings.gradle` is under user control, then it should not crash the tool (from the new code in this PR) if it is malformed, and there should be unit tests of that.